cot/src/test.rs

```diff
@@ -1234,6 +1239,275 @@ impl DynMigration for TestMigration {
     }
 }
 
+/// A utility for running entire projects in end-to-end tests.
+///
+/// This is useful for testing the full stack of a project, including the
+/// database, the router, the auth, etc. The server is running in the same
+/// process as the test by running a background async task.
+///
+///  This can be used to test the entire project by sending real requests to the
+/// server, possibly using libraries such as
+/// - [`reqwest`](https://docs.rs/reqwest/latest/reqwest/) for HTTP requests
+/// - [`thirtyfour`](https://docs.rs/thirtyfour/latest/thirtyfour/) or [`fantoccini`](https://docs.rs/fantoccini/latest/fantoccini/)
+///   for browser automation
+///
+/// Note that you need to use [`cot::e2e_test`] to run this, not
+/// [`mod@cot::test`]. Remember to call [`TestServerRunning::close`] when you're
+/// done with the tests, as the server will not be stopped automatically.
+///
+/// # Examples
+///
+/// ```
+/// use cot::test::TestServer;
+///
+/// struct TestProject;
+/// impl cot::Project for TestProject {}
+///
+/// #[cot::e2e_test] // note this uses "e2e_test"!
+/// async fn test_server() -> cot::Result<()> {
+///     let server = TestServer::new(TestProject).start().await;
+///
+///     let url = server.url();
+///     // ...use the server URL to send requests to the server
+///
+///     server.close().await;
+///     Ok(())
+/// }
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TestServer<T> {
```

Comment:
What's the point of having `TestServer` and `TestServerRunning`? It seems like they could be merged without any loss to the functionality

Reply:
Type safety. `TestServer` should not be allowed to `start()` twice.

Reply:
Okay, I understand it now. But if that's the case, I think it might be worth it to rename those to `TestServerBuilder` and `TestServer` respectively.